### PR TITLE
fix: Calendar select can't switch

### LINKED ISSUE #15334: Calendar select cannot switch
https://github.com/ant-design/ant-design/pull/15303

Reproduce demo: https://ant.design/components/calendar/

### PR DESCRIPTION
**Pull request by orzyyyy**
First of all, thank you for your contribution! 😄

New feature please send pull request to feature branch, and rest to master branch.
Pull request will be merged after one of collaborators approve.
Please makes sure that these form are filled before submitting your pull request, thank you!

[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md)]

### 🤔 This is a ...

- [ ] New feature
- [x] Bug fix
- [ ] Site / document update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Branch merge
- [ ] Other (about what?)

### 👻 What's the background?

1. Describe the source of requirement, like related issue link.

    To close #15334

2. Describe the problem and the scenario.

### 💡 Solution

1. How to fix the problem, and list final API implementation and usage sample if that is an new feature.

2. GIF or snapshot should be provided if includes UI/interactive modification.

### 📝 Changelog description

> Describe changes from user side, and list all potential break changes or other risks.

    Fix Calendar can't switch.

    修复 Calendar 无法切换类型的问题。

### ☑️ Self Check before Merge

- [ ] Doc is updated/provided or not needed
- [ ] Demo is updated/provided or not needed
- [ ] TypeScript definition is updated/provided or not needed
- [ ] Changelog is provided or not needed

### CODE REVIEW

**Comment by afc163** [comment]
We should add some test cases..

**Comment by afc163** [comment]
所以，问题是出在哪里了？

**Comment by zy410419243** [comment]
`getDerivedStateFromProps` 里判断 props 里有 `mode` 就更新 state， 然后 defaultProps.mode 始终都是 `month`，也就是 Header 的 onChange 始终没法改变外部的 mode

**Comment by afc163** [comment]
话说这个组件用的人真少，居然都没人报 bug。

**Comment by zy410419243** [comment]
何不弄个问卷，或者直接在文档网页上弄个点击统计

不但方便路人看热闹，自己还能看到哪些组件该重点维护

File: components/calendar/__tests__/index.test.js [hunk 1]
```diff
@@ -161,4 +161,16 @@ describe('Calendar', () => {
...
+  it('switch should work correctly without prop mode', () => {
+    const onPanelChange = jest.fn();
+    const wrapper = mount(<Calendar onPanelChange={onPanelChange} />);
+    expect(wrapper.state().mode).toBe('month');
+    wrapper
+      .find('.ant-radio-button-input')
+      .at(1)
+      .simulate('change');
+    expect(onPanelChange).toBeCalled();
+    expect(onPanelChange.mock.calls[0][1]).toEqual('year');
```
Comment (afc163) [inline components/calendar/__tests__/index.test.js]: 最好检测一下面板内容是否切换了，直接 toMatchSnapshot 是一个好办法。

**Comment by afc163** [comment]
image

File: components/calendar/__tests__/index.test.js [hunk 2]
```diff
@@ -170,6 +170,7 @@ describe('Calendar', () => {
       .find('.ant-radio-button-input')
       .at(1)
       .simulate('change');
+    expect(wrapper).toMatchSnapshot();
```
Comment (afc163) [inline components/calendar/__tests__/index.test.js]: 切换前 `toMatchSnapshot` 一次，切换后再 `toMatchSnapshot` 一次，变化前后都留快照。

File: components/calendar/__tests__/index.test.js [hunk 3]
```diff
@@ -161,4 +161,18 @@ describe('Calendar', () => {
...
     expect(onPanelChange.mock.calls[0][0].month()).toEqual(date.month() - 1);
   });
+
+  it('switch should work correctly without prop mode', () => {
+    const onPanelChange = jest.fn();
+    const date = new Moment('1990-09-03');
+    const wrapper = mount(<Calendar onPanelChange={onPanelChange} value={date} />);
+    expect(wrapper.state().mode).toBe('month');
+    wrapper
+      .find('.ant-radio-button-input')
```
Comment (afc163) [inline components/calendar/__tests__/index.test.js]: `.ant-radio-button-input[value="year"]` 语义更准确一点。

**Comment by zombieJ** [comment]
> 话说这个组件用的人真少，居然都没人报 bug。

3.1.4.x 是好的：https://codesandbox.io/s/m9ol1o05nj

感觉迁移 lifecycle j经常遇到这类问题，合之前需要多点点 😄 

ref: https://github.com/ant-design/ant-design/commit/4601144f3fcdca90aba4267f57f70df6756e1052

**Review by ztplz — APPROVED** [review]

**Comment by zombieJ** [comment]
```diff
- const mode = type === 'date' ? 'month' : 'year';
- if (this.state.mode !== mode) {
-     this.setState({ mode });	
-     this.onPanelChange(this.state.value, mode);	
- }

+ setType = (mode: CalendarMode) => {
+   this.setState({ mode });
+   this.onPanelChange(this.state.value, mode);
+ }
```

感觉 type 也没什么用了，全部统一掉吧:
* mode => type
* setType  =>  setMode
* onHeaderTypeChange => onHeaderModeChange

**Pull request merged**